packages/subcommands/src/lib/utils/SubcommandsHandler.ts

```diff
@@ -0,0 +1,121 @@
+import type { Command } from '../structures/Command';
+import { SubcommandNotLoadedError } from '../errors/SubcommandNotLoadedError';
+import {
+	Args,
+	IPreconditionContainer,
+	PreCommandRunPayload,
+	SapphireClient,
+	Store,
+	PreconditionContainerArray,
+	PreconditionContainerSingle
+} from '@sapphire/framework';
+
+export type ResolveSubcommandPayload = Omit<PreCommandRunPayload, 'context'>;
+export interface ResolveSubcommandResult {
+	args: Args;
+	subcommand: Command | undefined;
+}
+
+export interface SubcommandsPluginOptions {
+	/**
+	 * Continue execution of parent command if command is disabled or error occurred
+	 * @since 1.0.0
+	 * @default false
+	 */
+	continueParent?: boolean;
+
+	/**
+	 * Run parent command when subcommand execution was ended
+	 * @since 1.0.0
+	 * @default false
+	 */
+	runParentAfterSubcommand?: boolean;
+
+	/**
+	 * Preconditions that will be overlap in parents by their subcommands
+	 * @since 1.0.0
+	 * @default []
+	 */
+	overlappedPreconditions?: string[];
+}
+
+export const enum SubcommandsEvents {
+	PreSubcommandRun = 'preSubcommandRun',
+	SubcommandDenied = 'subcommandDenied',
+	SubcommandAccepted = 'subcommandAccepted',
+	SubcommandRun = 'subcommandRun',
+	SubcommandSuccess = 'subcommandSuccess',
+	SubcommandError = 'subcommandError',
+	SubcommandFinish = 'subcommandFinish'
+}
+
+export class SubcommandsHandler {
+	// eslint-disable-next-line @typescript-eslint/naming-convention
+	public readonly subcommandsPluginOptions: Required<SubcommandsPluginOptions>;
+
+	public constructor(private readonly client: SapphireClient, options: SubcommandsPluginOptions) {
+		this.subcommandsPluginOptions = {
+			continueParent: options?.continueParent ?? false,
+			runParentAfterSubcommand: options?.runParentAfterSubcommand ?? false,
+			overlappedPreconditions: options?.overlappedPreconditions ?? []
+		};
+	}
+
+	public async resolveSubcommand({ command, message, parameters }: ResolveSubcommandPayload): Promise<ResolveSubcommandResult> {
+		const args = await command.preParse(message, parameters);
+		const store = command.store as Store<Command>;
+		let subcommandFromArgs = await args.pickResult('string');
+		let iterableCommand = command;
+		do {
+			if (!subcommandFromArgs.success || !iterableCommand.subcommands) break;
+
+			const soughtName = subcommandFromArgs.value.toLowerCase();
+
+			const soughtSubcommand = store.find((c) => c.name === soughtName || c.aliases.includes(soughtName));
+			if (!soughtSubcommand || soughtSubcommand.parentName !== iterableCommand.name) break;
+			iterableCommand = soughtSubcommand;
+		} while ((subcommandFromArgs = await args.pickResult('string')));
+
+		return { subcommand: iterableCommand === command ? undefined : iterableCommand, args };
+	}
+
+	public buildSubcommandsTree() {
+		const { commands } = this.client;
```

Comment:
Would've been easier to extend the command store, however, the store may not always exist or be under `client.commands` as stated later in this review. Consider adding a parameter perhaps, or changing the plugin to extend the store.

Reply:
I decide to not extending store, because subcommands is the same as Command and recognize by framework the same. No need to create another unnecessary store, i thought

Reply:
Argument added, client field removed from SubcommandsHandler